File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/model/BaseModelResolver.java
package org.opencds.cqf.cql.engine.model;

import org.opencds.cqf.cql.engine.exception.InvalidCast;

public abstract class BaseModelResolver implements ModelResolver {
    public Boolean is(Object value, Class<?> type) {
        if (value == null) {
            return null;
        }

        return type.isAssignableFrom(value.getClass());
    }

    public Object as(Object value, Class<?> type, boolean isStrict) {
        if (value == null) {
            return null;
        }

        if (type.isAssignableFrom(value.getClass())) {
            return value;
        }

        if (isStrict) {
            throw new InvalidCast(String.format(
                    "Cannot cast a value of type %s as %s.", value.getClass().getName(), type.getName()));
        }

        return null;
    }
}

File: Src/java/engine/src/main/kotlin/org/opencds/cqf/cql/engine/model/BaseModelResolver.kt
package org.opencds.cqf.cql.engine.model

import org.opencds.cqf.cql.engine.exception.InvalidCast

abstract class BaseModelResolver : ModelResolver {
    override fun `is`(value: Any?, type: Class<*>?): Boolean? {
        if (value == null) {
            return null
        }

        return type!!.isAssignableFrom(value.javaClass)
    }

    override fun `as`(value: Any?, type: Class<*>?, isStrict: Boolean): Any? {
        if (value == null) {
            return null
        }

        if (type!!.isAssignableFrom(value.javaClass)) {
            return value
        }

        if (isStrict) {
            throw InvalidCast(
                "Cannot cast a value of type ${value.javaClass.getName()} as ${type.getName()}."
            )
        }

        return null
    }
}
